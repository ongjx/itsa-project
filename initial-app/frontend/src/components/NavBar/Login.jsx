import React, { useState } from 'react';
import { BrowserRouter as Router, Switch, Link, Route, useHistory } from "react-router-dom";
import { Button, Form } from 'react-bootstrap';
import axios from "axios";
import Alert from '@material-ui/lab/Alert';
import cookie from 'react-cookies'

const Login = ({ loginData }) => {
    const defaultValues = {
        email: "",
        password: ""
    };

    const [formValues, setFormValues] = useState(defaultValues)
    const history = useHistory();
    const [message, setMessage] = useState("");
    const [severity, setSeverity] = useState("");
    const [response, setResponse] = useState([])
    const handleChange = (e) => {
        var { name, value } = e.target;
        setFormValues({
            ...formValues,
            [name]: value,
        });
    }

    const expires = new Date()
    expires.setDate(Date.now() + 1000 * 60 * 60 * 24 * 14)

    const handleSubmit = async (e) => {
        e.preventDefault();
        setMessage("")
        setSeverity("")
        const fetchURL = "https://api.ascendahotels.me/backend/login";
        if(formValues.email=="" || formValues.password==""){
            setMessage("Please fill in email/password")
            setSeverity("error")
        }else{
            try{
                await axios.post(fetchURL, {
                    customerEmail: formValues.email,
                    customerPassword: formValues.password
                }).then((res) => {
                    setResponse(res)
                    setMessage("Login Success")
                    setSeverity("success")
                    history.push({
                        pathname: `/profile`,
                        state: {
                            loginData: formValues
                        }
                    })
                    cookie.save("Authorization", "yourSecretT0k4n", {
                        path: "/",
                        expires,
                        maxAge: 1000,
                        // domain: "https://*.ascendahotels.me",
                        // secure: true,
                        // httpOnly: true
                    })
                });
            }catch(err){
                console.log(err)
                setMessage("Wrong Credentials")
                setSeverity("error")
            }finally{
                if(response.length==0){
                    setMessage("Wrong Credentials")
                    setSeverity("error")
                }
            }
        }
    }
    return (
        <div className='text-left p-4'>
            <h2 className='mb-4'>
                Login
            </h2>
            <Form onSubmit={handleSubmit}>
                <Form.Group controlId="email">
                    <Form.Label>Email</Form.Label>
                    <Form.Control type="text" name="email" placeholder="email" onChange={handleChange} />
                </Form.Group>

                <Form.Group controlId="password">
                    <Form.Label>Password</Form.Label>
                    <Form.Control type="password" name="password" placeholder="Password" onChange={handleChange} />
                </Form.Group>

                <Button variant="primary" type="submit" style={{ 'width': '100%' }}>
                    Login
                </Button>

            </Form>
            {message? <Alert severity={severity}>{message}</Alert>: <></>}
        </div>
    )
}
export default Login