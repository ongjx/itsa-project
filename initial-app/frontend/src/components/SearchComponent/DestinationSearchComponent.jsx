import React, { useEffect, useState } from 'react';
import FuzzySearch from 'react-fuzzy';
import { Button, Form } from 'react-bootstrap';
import moment from 'moment'
import {  useHistory, useLocation } from 'react-router-dom';
import HotelCardComponent from "../Cards/HotelCardComponent"
import axios from 'axios'
export default  function DestinationSearchComponent({ destinationsData, hotelsData }) {
    let history = useHistory();
    let location = useLocation();
    const [destinations, setDestinations] = useState(destinationsData)
    const [hotels, setHotels] = useState(hotelsData)
    const [filteredHotels, setFilteredHotels] = useState([])
    useEffect(() => {
        setDestinations(destinationsData)
        setHotels(hotelsData)
    }, [destinationsData, hotelsData])

    const defaultValues = {
        destination: "",
        hotel: "",
        checkInDate: "",
        checkOutDate: "",
        numGuest: 1,
        numRoom: 1,
    };
    const [formValues, setFormValues] = useState(defaultValues)
    const [items, setItems] = useState()

    const setSelectedDestination = (selectedDestination) => {
        setFormValues({
            ...formValues,
            'destination': selectedDestination
        })
        // on select destination, filter hotels
        getFilteredHotels(selectedDestination.destination)
    }
    const getFilteredHotels = (destination) => {
        return axios.get(`https://api.ascendahotels.me/hotels?destination_id=${destination}`).then(res => {
            const data = res.data
            const hotels = data.map(hotel => {return {id: hotel.uid, hotel: hotel.hotel}})
            setFilteredHotels(hotels)
        })
    }

    const setSelectedHotel = (selectedHotel) => {
        setFormValues({
            ...formValues,
            'hotel': selectedHotel,
        });
    }
    const handleChange = (e) => {
        var { name, value } = e.target;

        if (name == 'numRoom' || name == 'numGuest') {
            value = parseInt(value.split(' ')[0])
        }

        setFormValues({
            ...formValues,
            [name]: value,
        });
    }

    const detectChange = (e, type) => { // handle change in destination value
        if (e.target.value === "" && type==='destination') {
            setFilteredHotels([]);
        }
        if (e.target.value === "" && type==='hotel'){
            setFormValues({
                ...formValues,
                'hotel': ''
            })
        }
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        const destination = formValues.destination.destination
        const destinationId = formValues.destination.id
        const hotel = formValues.hotel.id
        const hotel_name = formValues.hotel.hotel
        const checkInDate = formValues.checkInDate
        const checkOutDate = formValues.checkOutDate
        const guest = formValues.numGuest
        const room = formValues.numRoom

        if (hotel && destination) {
            const fetchURL = `https://api.ascendahotels.me/hotels/${hotel}`;
            axios.get(fetchURL).then
            (res => {
                history.push({
                    pathname: `/hotel-details`,
                    state: {
                        pageBefore: '/',
                        hotelObject: res.data,
                        isBookable: true,
                        formValues:formValues,
                        destination: destination,
                        destinationId: destinationId,
                        checkInDate: checkInDate,
                        checkOutDate: checkOutDate,
                        guest: guest,
                        numroom: room
                    }
                })
            })
        }
        else if (hotel === undefined) { //only dest selected; no hotel was selected
            const fetchURL = `https://api.ascendahotels.me/hotels/info?destination_id=${destinationId}&checkin=${checkInDate}&checkout=${checkOutDate}&guest=${guest}`; // TODO: CHANGE THIS TO DYNAMIC URL
            axios.get(fetchURL).then(res => {
                setItems(res.data)
                history.push({
                    pathname: `/all-hotels`,
                    state: {
                        pageBefore: '/',
                        rooms: res.data,
                        destination: destination,
                        destinationId: destinationId,
                        checkInDate: checkInDate,
                        checkOutDate: checkOutDate,
                        guest: guest,
                        numroom: room
                    }
                })
            })
        }
    }
    const minDate = moment(new Date()).format("YYYY-MM-DD")
    return (
        <div className='text-left p-4'>
            <h2 className='mb-4'>
                Search for Hotels
            </h2>
            <Form onSubmit={handleSubmit}>
                <Form.Group controlId="destination"  onChange={(e) => detectChange(e,"destination")}>
                    <Form.Label>Destination</Form.Label>
                    <FuzzySearch
                        className='shadow-none' inputWrapperStyle={{ 'boxShadow': '' }} inputStyle={{ 'border': '1px solid #ced4da' }}
                        list={destinations ? destinations : []}
                        width={'100%'}
                        keys={['destination', 'id']}
                        keyForDisplayName={'destination'}
                        threshold={0.4}
                        placeholder={"Search Destination"}
                        onSelect={(newSelectedItem) => {
                            // Local state setter defined elsewhere
                            setSelectedDestination(newSelectedItem)
                        }}
                        shouldShowDropdownAtStart={false}
                    />
                </Form.Group>
                {
                    filteredHotels.length == 0 ? <></> :

                <Form.Group controlId="hotel" onChange={(e) => detectChange(e, "hotel")}>
                    <Form.Label>Hotel</Form.Label>
                    <FuzzySearch
                        className='shadow-none' inputWrapperStyle={{ 'boxShadow': '' }} inputStyle={{ 'border': '1px solid #ced4da' }}
                        list={filteredHotels}
                        width={'100%'}
                        keys={['id', 'hotel']}
                        keyForDisplayName={'hotel'}
                        threshold={0.4}
                        placeholder={"Search Hotel"}
                        onSelect={(newSelectedItem) => {
                            // Local state setter defined elsewhere
                            setSelectedHotel(newSelectedItem)
                        }}
                        shouldShowDropdownAtStart={true}
                    />
                </Form.Group>
                }
                <Form.Group>
                    <Form.Label>Check In Date</Form.Label>
                    <Form.Control name='checkInDate' type='date' min={minDate} onChange={handleChange} />
                </Form.Group>
                <Form.Group>
                    <Form.Label>Check Out Date</Form.Label>
                    <Form.Control name='checkOutDate' type='date' min={minDate} onChange={handleChange} />
                </Form.Group>
                <Form.Group controlId="formBasicCheckbox">
                    <Form.Label>Number of Guests</Form.Label>
                    <Form.Control as="select" label="guests" name='numGuest' onChange={handleChange}>
                        <option>1 Adult</option>
                        <option>2 Adult</option>
                        <option>3 Adult</option>
                        <option>4 Adult</option>
                        <option>5 Adult</option>
                    </Form.Control>
                </Form.Group>
                <Form.Group controlId="formBasicCheckbox">
                    <Form.Label>Rooms</Form.Label>
                    <Form.Control as="select" label="rooms" name='numRoom' onChange={handleChange}>
                        <option>1 Room</option>
                        <option>2 Room</option>
                        <option>3 Room</option>
                        <option>4 Room</option>
                        <option>5 Room</option>
                    </Form.Control>
                </Form.Group>
                <Button variant="outline-warning" type="submit" style={{ 'width': '100%' }}>
                    Check Availability
                </Button>

                <div className='mt-4'>
                    {/* <HotelCardComponent image="http://photos.hotelbeds.com/giata/bigger/06/069630/069630a_hb_ro_003.jpg" name="Hotel" price="$xx"></HotelCardComponent> */}
                </div>
            </Form>
        </div>
    )
}

