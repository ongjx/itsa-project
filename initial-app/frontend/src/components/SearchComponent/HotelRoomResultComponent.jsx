import React, { useState, useEffect } from 'react';
import axios from 'axios'
import HotelCardComponent from "../Cards/HotelCardComponent"
import Grid from '@material-ui/core/Grid';
import { useHistory, useLocation } from 'react-router-dom';
import HotelInfoComponent from '../HotelInfoComponent/HotelInfoComponent';
import BackButtonComponent from '../BackButtonComponent';

const HotelRoomResultComponent = ({ callbackFn, hotelRoomId }) => {
    const history = useHistory();
    const state = useLocation();
    const [rooms, setRooms] = useState(state.state.rooms);
    const [prices, setPrices] = useState();

    useEffect(() => {
        sendWithRetry();
        rooms.sort((a,b) => a.searchRank < b.searchRank)
    }, []);

    const MAX_RETRY = 1;
    let currentRetry = 0;

    const errorHandler = () => {
        if (currentRetry < MAX_RETRY) {
            currentRetry++;
            setTimeout(() => {
                sendWithRetry();
            },3000)
        }
    }

    const sendWithRetry = () => {
        const destinationID = state.state.destinationId;
        const checkInDate = state.state.checkInDate;
        const checkOutDate = state.state.checkOutDate;
        const guest = state.state.guest;
        const fetchPriceURL = `https://api.ascendahotels.me/prices?destination_id=${destinationID}&checkin=${checkInDate}&checkout=${checkOutDate}&guests=${guest}`;
        axios.get(fetchPriceURL).then((res) => {
            const newPrices = res.data.hotels
            setPrices(newPrices)
            const isComplete = res.data.completed
            if (!isComplete) {
                errorHandler()
            }
        })
    }

    return (
        <>
            <BackButtonComponent pageBefore={state.pathname}/>
            {
                rooms && prices ?
                    <Grid container spacing={3} direction="row" justify="center">
                        {
                            prices && rooms ? rooms.map((item) => {
                                return (
                                    <Grid item xs={12} sm={6} md={4}>
                                        <HotelCardComponent hotelObject={item} priceObject={prices ? prices.filter(priceItem => priceItem.id === item.id) : []}/>
                                    </Grid>
                                )
                            }) : <></>
                        }
                    </Grid>
                    : <></>
            }

        </>
    )
}

export default HotelRoomResultComponent