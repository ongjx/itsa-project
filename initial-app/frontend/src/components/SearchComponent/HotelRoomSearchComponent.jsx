import React, { useState, useEffect } from "react";
import { useHistory, useLocation } from "react-router-dom";
import axios from "axios";
import HotelCardComponent from "../Cards/HotelCardComponent";
import Grid from "@material-ui/core/Grid";
import BackButtonComponent from "../BackButtonComponent";
import { css } from "@emotion/core";
import ClipLoader from "react-spinners/ClipLoader";

const override = css`
  display: block;
  margin: 0 auto;
  border-color: blue;
  radius: 3;
`;

export default function HotelRoomSearchComponent({ hotelsData }) {
  const history = useHistory();
  const state = useLocation();
  const [rooms, setRooms] = useState(state.state.rooms);
  const [prices, setPrices] = useState(state.state.prices);
  const [loading, setLoading] = useState(true);
  const [color, setColor] = useState("#ffffff");

  useEffect(() => {
    if (!rooms) {
      setRooms(state.state.rooms);
      if (rooms && rooms.length > 0) {
        rooms.sort((a, b) => a.searchRank < b.searchRank);
      }
    }
    if (!prices) {
      sendWithRetry();
    }else{
      setLoading(false)
    }
  }, []);

  const sendWithRetry = () => {
    const destinationID = state.state.destinationId;
    const checkInDate = state.state.checkInDate;
    const checkOutDate = state.state.checkOutDate;
    const guest = state.state.guest;
    const fetchPriceURL = `https://api.ascendahotels.me/prices?destination_id=${destinationID}&checkin=${checkInDate}&checkout=${checkOutDate}&guest=${guest}`;

    axios.get(fetchPriceURL).then((res) => {
      const newPrices = res.data.hotels;
      setPrices(newPrices);
      setLoading(false)
    });
  };

  return (
    <>
      <BackButtonComponent pageBefore={"/"} />
      {loading? (
          <div style={{ paddingBottom: 30 }}>
            Loading
            <ClipLoader color={color} loading={true} css={override} size={40}/>
          </div>
        ) : (
          <></>
        )}
      {rooms && prices ? (
        <Grid container spacing={3} direction="row" justify="center">
          {prices && rooms ? (
            rooms.map((item) => {
              return (
                <Grid item xs={12} sm={6} md={4}>
                  <HotelCardComponent
                    hotelObject={item}
                    priceObject={
                      prices
                        ? prices.filter((priceItem) => priceItem.id === item.id)
                        : []
                    }
                    allPrices={prices}
                  />
                </Grid>
              );
            })
          ) : (
            <></>
          )}
        </Grid>
      ) : (
        <></>
      )}
    </>
  );
}
