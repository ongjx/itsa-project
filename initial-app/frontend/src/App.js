import './App.css';
import AppRoutes from './AppRoutes';
import NavBarComponent from './components/NavBar/NavBarComponent';
import SearchComponent from './components/SearchComponent/SearchComponent';
import { BrowserRouter } from "react-router-dom";
function App() {
  return (
    <div className="App">
      <BrowserRouter>
      <NavBarComponent/>
      <div className='container'>
        <AppRoutes />
      </div>
      </BrowserRouter>
    </div>
  );
}

export default App;
