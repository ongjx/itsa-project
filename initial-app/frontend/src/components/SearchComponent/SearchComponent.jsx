import React, { useState, useEffect } from 'react';
import DestinationSearchComponent from './DestinationSearchComponent';
import axios from 'axios'
const SearchComponent = () => {
    const [selectedDestination, setSelectedDestination] = useState('');
    const [filteredHotels, setFilteredHotels] = useState([]);
    const [destinationsData, setDestinationsData] = useState();
    const [hotelsData, setHotelsData] = useState();

    const [cache, setCache] = useState();

    const retrieveDestinationData = () => {
        const cachedHits = localStorage.getItem("destinationData")
        let expiration = localStorage.getItem("expiration")
        let currTime = new Date().getTime()
        if (cachedHits && expiration && expiration > currTime) {
            setCache(cachedHits)
            getDestinations(cachedHits)
        } else {
            axios.get("https://api.ascendahotels.me/destinations").then(res => {
                const cache = JSON.stringify(res.data)
                localStorage.setItem("destinationData", cache)
                expiration = currTime + 86400000
                localStorage.setItem("expiration", expiration)
                setCache(cache)
                getDestinations(cache)
            })
        }
    }
    const getDestinations = (cache) => {
        const destinations = JSON.parse(cache).map((destinationObj => Object.entries(destinationObj).map((destination) => { return {destination: destination[0], id: destination[1]}}))).reduce((memo, it) => (memo.concat(it)), [])
        setDestinationsData(destinations)
    }

    const getHotels = (cache) => {
        const hotels = Object.values(JSON.parse(cache)).map(hotels => Object.entries(hotels.hotels).map(hotel => {return {id: hotels.primary_destination_id, hotel: hotel[0], hotel_id:hotel[1]}})).reduce((memo, it) => (memo.concat(it)), [])
        setHotelsData(hotels)
    }
    useEffect(() => {
        retrieveDestinationData()
    }, [])

    return (
        <div>
            <DestinationSearchComponent destinationsData={destinationsData} hotelsData={hotelsData}/>
        </div>
    )

}
export default SearchComponent