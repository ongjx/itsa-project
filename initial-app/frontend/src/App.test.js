import React from 'react'
import {render, cleanup} from '@testing-library/react'
import App from './App'
import SearchComponent from "./components/SearchComponent/SearchComponent";
import Login from "./components/NavBar/Login";

 afterEach(cleanup)
 
 it('should take a snapshot of Search Component', () => {
    const { asFragment } = render(<Login />)
    
    expect(asFragment(<Login />)).toMatchSnapshot()
   })

