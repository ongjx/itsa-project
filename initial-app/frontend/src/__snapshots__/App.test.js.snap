// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot of Search Component 1`] = `
<DocumentFragment>
  <div
    class="text-left p-4"
  >
    <h2
      class="mb-4"
    >
      Login
    </h2>
    <form
      class=""
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="email"
        >
          Email
        </label>
        <input
          class="form-control"
          id="email"
          name="email"
          placeholder="email"
          type="text"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="password"
        >
          Password
        </label>
        <input
          class="form-control"
          id="password"
          name="password"
          placeholder="Password"
          type="password"
        />
      </div>
      <button
        class="btn btn-primary"
        style="width: 100%;"
        type="submit"
      >
        Login
      </button>
    </form>
  </div>
</DocumentFragment>
`;
